sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2458,6 +2516,192 @@ checkDevSupportDeviceRequirements(const device &Dev,
     }
   }
 
+  // TODO: remove checks for CUDA and HIP from if-statement below when runtime
+  // query for them in matrix_combinations is implemented
+  if (JointMatrixPropIt &&
+      (Dev.get_backend() != sycl::backend::ext_oneapi_cuda) &&
+      (Dev.get_backend() != sycl::backend::ext_oneapi_hip)) {
+    std::vector<ext::oneapi::experimental::matrix::combination> Combinations =
+        Dev.get_info<
+            ext::oneapi::experimental::info::device::matrix_combinations>();
+
+    if (Combinations.empty())
+      return sycl::exception(errc::kernel_not_supported,
```

Comment:
I wonder, why does this code compile?

`sycl::exception` constructors are [defined](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:exception.class) to accept either `std::error_code` or `int` as the first argument, but here we are passing `sycl::errc` of unspecified (in the spec) type. Even if it is specified as `int` in our implementation, it should be defined as an `enum class`, which would require an explicit `static_cast` to `int`.

Am I missing something here, or do we have some extra non-standard `sycl::exception` constructors?
The right way to go is through `std::error_code make_error_code(errc e) noexcept;` if I understand correctly

Reply:
Looks like this is because of 
https://github.com/intel/llvm/blob/8cb3343f6435975d1e7ef5ee94a7fd0a387b7df4/sycl/include/sycl/exception.hpp#L336

https://en.cppreference.com/w/cpp/error/error_code/is_error_code_enum

Should we be more strict to the SYCL2020 spec here?

Reply:
> Should we be more strict to the SYCL2020 spec here?

Can you clarify what should be more strict in the spec?

Reply:
> > Should we be more strict to the SYCL2020 spec here?
> 
> Can you clarify what should be more strict in the spec?

When I wrote the message above, I didn't look into the spec and I thought that specialization of `std::is_error_code_enum` is not according to the spec, so the question was - should our implementation be more strict (to align with the spec), but this specialization is in the spec, so no question from my side now - the code above is valid and in accordance to the SYCL2020 spec.

Reply:
Oh, I wasn't aware of `std::is_error_code_enum` functionality and hence my misunderstanding of why we can pass `sycl::errc` directly to `std::error_code` argument. With that knowledge in mind, we can actually refactor plenty of places in our runtime to get rid of `make_error_code` calls and make the code shorter